cmd/geth: implement dev mode for post-merge

jwasinger opened the pull request:
Opening this as a draft to get eyes on it in its current state, and get feedback if there are obvious things that should be changed.

Currently works with the command: `./build/bin/geth --dev`

holiman commented:
> Currently works with the command: `./build/bin/geth --http --http.api "engine" --nodiscover --authrpc.jwtsecret $(pwd)/jwt_secret --dev`

Haven't looked at the code, but why would one want to pass jwtsecret? Shouldn't it just be `geth --dev` and done?

jwasinger commented:
@holiman yup.  That's something I've noted in todo list above.  Trying to address those today.

s1na commented on clmock/clmock.go:
```diff
@@ -0,0 +1,156 @@
+			break
+		case curTime := <-ticker.C:
+			if curTime.After(lastBlockTime.Add(blockPeriod)) {
+				safeHead, err := engine_api.GetHeaderByTag(c.ctx, "safe")
```
I feel like these "safe" and "finalized" fetches are not needed. There's no other proposer so it should be able to keep track of these blocks (they are already tracked in `curForkchoiceState`)?

s1na commented on clmock/clmock.go:
```diff
@@ -0,0 +1,156 @@
+
+	for {
+		select {
+		case _ = <-c.ctx.Done():
```
```suggestion
		case <-c.ctx.Done():
```

lightclient commented on clmock/clmock.go:
```diff
@@ -0,0 +1,156 @@
+	var prevRandaoVal common.Hash
+	var suggestedFeeRecipient common.Address
+
+	engine_api := engineAPI{}
```
```suggestion
	engine := engineAPI{}
```
snake case

jwasinger commented on clmock/engine_api.go:
```diff
@@ -0,0 +1,87 @@
+func (e *engineAPI) Connect(ctx context.Context, endpoint string) error {
+	// TODO don't hardcode these here
+	var testSecret = [32]byte{94, 111, 36, 109, 245, 74, 43, 72, 202, 33, 205, 86, 199, 174, 186, 77, 165, 99, 13, 225, 149, 121, 125, 249, 128, 109, 219, 163, 224, 176, 46, 233}
+	var testEndpoint = "http://127.0.0.1:8551"
```
reworked this code to access the engine API methods directly and not via rpc calls.

jwasinger commented:
Okay.  This is getting closer.

Two items still need to be addressed:
* configurability of dev mode gas limit (override genesis configuration if `--dev.gaslimit` flag is present and there is no pre-existing chain)
* currently panics if started up using an incompatible datadir.

gballet commented on clmock/clmock.go:
```diff
@@ -0,0 +1,171 @@
+	// how do we sync node shutdown with this separate go-routine?
+	// does it matter?  the worst that can happen is we get some weird error messages on node shutdown that might throw users off
+	ticker := time.NewTicker(time.Millisecond * 500)
+	blockPeriod := time.Second * 10 // hard-coded fast block period for testing purposes
```
I suggest that you make it a configurable switch with a default of 12.

gballet replied:
Isn't it why you created `DeveloperModeConfig` in the first place?

gballet commented on core/genesis.go:
```diff
@@ -568,19 +568,17 @@ func DefaultSepoliaGenesisBlock() *Genesis {
 	// Override the default period to the user requested one
-	config := *params.AllCliqueProtocolChanges
-	config.Clique = &params.CliqueConfig{
+	config := *params.AllDevChainProtocolChanges
```
not sure why you remove `AllCliqueProtocolChanges` here?

jwasinger replied:
Because developer mode no longer uses Clique for its consensus engine.

gballet commented on core/genesis.go:
```diff
@@ -568,19 +568,15 @@ func DefaultSepoliaGenesisBlock() *Genesis {
+		//ExtraData:  []byte{},
 		GasLimit:   gasLimit,
 		BaseFee:    big.NewInt(params.InitialBaseFee),
-		Difficulty: big.NewInt(1),
```
it was very difficult genecizing this genesis block

holiman replied:
The difficulty formula is undefined for difficulties `<131072`. There is a minimum difficulty, and below that, the spec can be interpreted in two different ways, depending on at what exact step one decides to apply the minimum threshold.

gballet commented on params/config.go:
```diff
@@ -299,8 +318,9 @@ type ChainConfig struct {
-	Clique *CliqueConfig `json:"clique,omitempty"`
+	Ethash *EthashConfig        `json:"ethash,omitempty"`
+	Clique *CliqueConfig        `json:"clique,omitempty"`
+	Dev    *DeveloperModeConfig `json:"dev,omitempty"`
```
Just my 2 cents here: I feel this is overkill, because you can just pass it as a command line option, and use that parameter when instantiating the `CLMock`. So you could reduce the footprint of that PR by not adding more config here, since it's really only used in one place.

jwasinger replied:
Yes but it must be persisted in the case where dev mode is not being used ephemerally.  Hence I add `DeveloperModeConfig` so that the period will be stored.

jwasinger replied:
When you specify `--datadir` and `--dev` together it persists the chain.  if the datadir is omitted, the chain is ephemeral.

lightclient replied:
I also feel this is overkill? Is there ever really a use case to have a dev mode genesis? `ChainConfig` is used quite a bit and so we should be confident in the addition here.

rjl493456442 replied:
In `DeveloperModeConfig `, there is only a single field `period`. It's totally fine to use another value in the next restart I think. We can have a default value for it, e.g. 12s or so, and users can change it by CLI flag.

gballet reviewed (commented):
Left a few questions

MariusVanDerWijden commented on clmock/clmock.go:
```diff
@@ -0,0 +1,171 @@
+func NewCLMock(stack *node.Node, backend ethapi.Backend) *CLMock {
+	c := CLMock{}
+	c.stack = stack
+	c.backend = backend
```
NVM I have some other ideas

jwasinger commented:
Alright, I've addressed the remaining issues that I listed above.  Adding a few unit tests and then hopefully, this should be ready for merge.

jwasinger commented:
It occurs to me that it probably makes sense to implement a way to shim-in support for withdrawals.  I'll try to think about how this should be supported.

lightclient commented on clmock/clmock.go:
```diff
@@ -0,0 +1,222 @@
+		HeadBlockHash:      header.Hash(),
+		SafeBlockHash:      header.Hash(),
+		FinalizedBlockHash: header.Hash(),
+	}
```
should probably just put all these in a `var` block - also no need to forward declare `curForkchoiceState`

lightclient commented on clmock/clmock.go:
```diff
@@ -0,0 +1,222 @@
+	}
+
+	// if genesis block, send forkchoiceUpdated to trigger transition to PoS
+	if header.Number.Cmp(big.NewInt(0)) == 0 {
```
```suggestion
	if header.Number.BitLen() == 0 {
```

lightclient commented on clmock/clmock.go:
```diff
@@ -0,0 +1,222 @@
+					SuggestedFeeRecipient: feeRecipient,
+				}
+
+				payloadAttr.Withdrawals = c.popWithdrawals()
```
this can be within the `payloadAttr` definition

lightclient commented on clmock/clmock.go:
```diff
@@ -0,0 +1,222 @@
+
+				// trigger block building (via forkchoiceupdated)
+				fcState, err := engineAPI.ForkchoiceUpdatedV2(curForkchoiceState, payloadAttr)
+
```
```suggestion
```

lightclient commented on clmock/clmock.go:
```diff
@@ -0,0 +1,222 @@
+				payloadTimeout := time.NewTimer(12 * time.Second)
+				// interval to poll the pending state to detect if transactions have arrived, and proceed if they have
+				// (or if there are pending withdrawals to include)
+				buildTicker := time.NewTicker(100 * time.Millisecond)
```
var block definition for the above 4 vars

lightclient commented on clmock/clmock.go:
```diff
@@ -0,0 +1,222 @@
+	withdrawals       []*types.Withdrawal
+	feeRecipient      common.Address
+	feeRecipientMutex sync.Mutex
+	nextWithdrawalIdx uint64
```
Seems like you could abstract this a bit into another type `withdrawalGenerator` and inside you could have this var as `next`.

jwasinger replied:
Right now, I just don't see how abstracting this into a separate type changes much here (other than adding some boilerplate).  All that's happening is that a counter is being incremented every time a new withdrawal is received, and that this counter is used to validate incoming withdrawals.

Now if validation of withdrawals in this code changes and becomes more complicated, I could see value in separating that logic into a separate type.

lightclient replied:
IMO it's more about naming convertions. `nextWithdrawalIdx`, `popWithdrawals` etc are pretty long, so I think a way to simplify a lot of the code here would be to make it it's own object. I was thinking something more along these lines: https://play.golang.com/p/Rbt6XQ8gpv_H

lightclient commented on clmock/clmock.go:
```diff
@@ -0,0 +1,222 @@
+		return fmt.Errorf("withdrawal has index (%d) less than or equal to latest received withdrawal index (%d)", w.Index, c.nextWithdrawalIdx-1)
+	}
+	c.nextWithdrawalIdx = w.Index + 1
+	c.withdrawals = append(c.withdrawals, &w)
```
i think that these two lines need to be done with the mutex? otherwise might be a race between the api and building blocks?

jwasinger replied:
I added a mutex lock for the entire function so that there isn't a race when reading the value of `c.nextWithdrawalIdx` above.

lightclient commented on clmock/api.go:
```diff
@@ -0,0 +1,22 @@
+package clmock
```
Might not even need a package, I think it could live happily inside `cmd/geth` also.

jwasinger commented:
Okay, I've addressed most of the feedback that @lightclient has provided (thanks for that!).  The only remaining question is where to put the `simulated_beacon` package?

It can either live under `internal` or in `cmd/geth` as suggested by @lightclient .

lightclient commented:
I think I lean towards just putting the files into `cmd/geth` and not adding a new package. It's relatively lightweight and only ever instantiated in `cmd/geth/config.go`. If it is a package though, you should [drop the underscore](https://go.dev/blog/package-names) in the name.

holiman commented:
Some UX quirks... 
It opens port `8551` -- which seems unnecessary ? 
```
INFO [06-13|11:56:39.169] IPC endpoint opened                      url=/tmp/geth.ipc
INFO [06-13|11:56:39.171] WebSocket enabled                        url=ws://127.0.0.1:8551
INFO [06-13|11:56:39.171] HTTP server started                      endpoint=127.0.0.1:8551 auth=true prefix= cors=localhost vhosts=localhost
```

When running it, it keep spitting out this: 
```
INFO [06-13|11:59:40.087] Starting work on payload                 id=0x1a411adabba04e97
INFO [06-13|11:59:40.087] Updated payload                          id=0x1a411adabba04e97 number=1 hash=3970ba..f2fc01 txs=0 gas=0 fees=0 root=e642b9..90f354 elapsed="263.675µs"
INFO [06-13|11:59:52.088] Stopping work on payload                 id=0x1a411adabba04e97 reason=timeout
INFO [06-13|11:59:52.089] Starting work on payload                 id=0xa0f6fb4c31da930b
INFO [06-13|11:59:52.089] Updated payload                          id=0xa0f6fb4c31da930b number=1 hash=051b8f..e98ddc txs=0 gas=0 fees=0 root=e642b9..90f354 elapsed="60.977µs"
INFO [06-13|12:00:04.089] Stopping work on payload                 id=0xa0f6fb4c31da930b reason=timeout
INFO [06-13|12:00:04.090] Starting work on payload                 id=0x7c0f68cd6ee81a44
INFO [06-13|12:00:04.090] Updated payload                          id=0x7c0f68cd6ee81a44 number=1 hash=5cfc3d..3ca163 txs=0 gas=0 fees=0 root=e642b9..90f354 elapsed="99.294µs"
INFO [06-13|12:00:16.091] Stopping work on payload                 id=0x7c0f68cd6ee81a44 reason=timeout
```
And that's while I'm not doing anything, just letting it idle. Seems unnecessary?

holiman commented on simulated_beacon/api.go:
```diff
@@ -0,0 +1,24 @@
+	simBeacon *SimulatedBeacon
+}
+
+func (api *API) AddWithdrawal(ctx context.Context, withdrawal *types.Withdrawal) error {
+	api.simBeacon.mu.Lock()
+	defer api.simBeacon.mu.Unlock()
+	return api.simBeacon.withdrawals.add(withdrawal)
+}
```
This is IMO a bit odd. I guess you wanted to split this up into a standalone API in order to make the distinction more clear about what is the public API-surface and what is the actual implementation? 

The thing which is a bit odd, IMO, is that you have this external wrapper which reaches in and locks/unlocks an internal mutex inside teh `simBeacon`. Either move the mutex to this `API`, or, probably better, move the locking so the `SimulatedBeacon` controls it's own locks.

holiman replied:
So this type would be a 'pure' wrapper: 
```suggestion
func (api *API) AddWithdrawal(ctx context.Context, withdrawal *types.Withdrawal) error {
	return api.simBeacon.withdrawals.add(withdrawal)
}
```
You could even (maybe?) turn this whole file into 
```golang
type API SimulatedBeacon

func (api *API) AddWithdrawal(ctx context.Context, withdrawal *types.Withdrawal) error {
	return api.withdrawals.add(withdrawal)
}

func (api *API) SetFeeRecipient(ctx context.Context, feeRecipient *common.Address) {
	api.feeRecipient = *feeRecipient
}
```
Depends on your preference I guess, but it's less code at least

The last change also requires a change here: 
```diff 
--- a/simulated_beacon/simulated_beacon.go
+++ b/simulated_beacon/simulated_beacon.go
@@ -216,7 +216,7 @@ func RegisterAPIs(stack *node.Node, c *SimulatedBeacon) {
        stack.RegisterAPIs([]rpc.API{
                {
                        Namespace: "dev",
-                       Service:   &API{c},
+                       Service:   (*API)(c),
                        Version:   "1.0",
                },
        })
```

jwasinger replied:
I've moved the mutex-handling for controlling access for fee recipient and the withdrawals list into `SimulatedBeacon`/`withdrawals` and pulled them out of the api.  This should make things a bit cleaner.

holiman commented on simulated_beacon/simulated_beacon.go:
```diff
@@ -0,0 +1,223 @@
+	}
+}
+
+// Start invokes the SimulatedBeacon life-cycle function in a goroutine
+func (c *SimulatedBeacon) Start() error {
+	c.ctx, c.cancel = context.WithCancel(context.Background())
+	go c.loop()
+	return nil
+}
+
+// Stop halts the SimulatedBeacon service
+func (c *SimulatedBeacon) Stop() error {
+	c.cancel()
+	return nil
+}
```
We usually do it like this 
```golang
func (c *SimulatedBeacon) Start() error {
	c.shutdownCh = make(chan struct{})
	go c.loop()
	return nil
}

// Stop halts the SimulatedBeacon service
func (c *SimulatedBeacon) Stop() error {
	close(c.shutdownCh)
	return nil
}
```
Then we only need to keep track of one thing -- not both a `ctx` and  `CancelFunc`

jwasinger replied:
Okay :+1: .  I think my use of a context here is a carry-over from when I was invoking RPC methods through the engine API external handlers (which require a context as a parameter).

holiman commented on simulated_beacon/simulated_beacon.go:
```diff
@@ -0,0 +1,236 @@
+		select {
+		case <-c.ctx.Done():
+			break
+		case curTime := <-ticker.C:
+			if curTime.Unix() > lastBlockTime.Add(c.period).Unix() {
+				c.mu.Lock()
+				feeRecipient := c.feeRecipient
+				pendingWithdrawals := c.withdrawals.pop()
+				c.mu.Unlock()
```
Also, maybe you could simplify the inner for-loop (and make some of those block-building messages go away, if you do the check on pending state earlier. Like this:

```golang 
		case curTime := <-ticker.C:
			if curTime.Unix() <= lastBlockTime.Add(c.period).Unix() {
				continue
			}
			c.mu.Lock()
			feeRecipient := c.feeRecipient
			pendingWithdrawals := c.withdrawals.pop()
			c.mu.Unlock()

			pendingHeader, err := c.eth.APIBackend.HeaderByNumber(context.Background(), rpc.PendingBlockNumber)
			if err != nil {
				log.Crit("failed to get pending block header", "err", err)
			}
			// don't build a block if we don't have pending txs or withdrawals
			if pendingHeader.TxHash == types.EmptyTxsHash && len(pendingWithdrawals) == 0 {
				continue
			}
			// Looks like it's time to build us a block!
```

holiman commented on simulated_beacon/simulated_beacon.go:
```diff
@@ -0,0 +1,236 @@
+					restartPayloadBuilding bool
+					// building a payload times out after SECONDS_PER_SLOT (12s on mainnet).
+					// trigger building a new payload if this amount of time elapses w/o any transactions or withdrawals
+					// having been received.
+					payloadTimeout = time.NewTimer(12 * time.Second)
+					// interval to poll the pending state to detect if transactions have arrived, and proceed if they have
+					// (or if there are pending withdrawals to include)
+					buildTicker = time.NewTicker(100 * time.Millisecond)
+				)
+				for {
+					select {
+					case <-buildTicker.C:
+						pendingHeader, err := c.eth.APIBackend.HeaderByNumber(context.Background(), rpc.PendingBlockNumber)
+						if err != nil {
+							log.Crit("failed to get pending block header", "err", err)
+						}
+						// don't build a block if we don't have pending txs or withdrawals
+						if pendingHeader.TxHash == types.EmptyTxsHash && len(payloadAttr.Withdrawals) == 0 {
+							continue
+						}
+						payload, err = engineAPI.GetPayloadV1(*fcState.PayloadID)
+						if err != nil {
+							log.Crit("error retrieving payload", "err", err)
+						}
+						// Don't build a block if it doesn't contain transactions or withdrawals.
+						// Somehow, txs can arrive, be detected by this routine, but be missed by the miner payload builder.
+						// So this last clause prevents empty blocks from being built.
+						if len(payload.Transactions) == 0 && len(payloadAttr.Withdrawals) == 0 {
+							restartPayloadBuilding = true
+						}
+					case <-payloadTimeout.C:
+						restartPayloadBuilding = true
+					case <-c.ctx.Done():
+						return
+					}
+					break
+				}
+				if restartPayloadBuilding {
+					continue
+				}
```
I think this entire block can be simplified into 
```suggestion
			time.Sleep(200 * time.Millisecond) // Give it some time to build
			if payload, err = engineAPI.GetPayloadV1(*fcState.PayloadID); err != nil {
				log.Crit("error retrieving payload", "err", err)
			}
			// Don't build a block if it doesn't contain transactions or withdrawals.
			if payload == nil || len(payload.Transactions) == 0 && len(payloadAttr.Withdrawals) == 0 {
				continue
			}
```

jwasinger replied:
This approach of having a static payload building time is something I discussed with @fjl .  It is simpler than the current approach.  However, the downside is that if users want to configure their dev chain with some incredibly large gas limit and execute very expensive transactions, they will not be included if they take too long to execute.

holiman replied:
Looks to me like your code is basically the same, but uses `100ms` instead of `200ms` ... ?

holiman commented on simulated_beacon/simulated_beacon.go:
```diff
@@ -0,0 +1,223 @@
+		case <-c.ctx.Done():
+			break
+		case curTime := <-ticker.C:
+			if curTime.Unix() > lastBlockTime.Add(c.period).Unix() {
```
Also, here's an alternative way to check if we have any transactions, without relying on the `pending` block: 
```golang
			if pendingTxs, _ := c.eth.APIBackend.TxPool().Stats(); pendingTxs == 0 && len(pendingWithdrawals) == 0 {
				continue
			}

```

holiman commented:
here's my slimmed down current version of the loop 

```golang
func (c *SimulatedBeacon) loop() {
	var (
		ticker             = time.NewTicker(time.Millisecond * 100)
		lastBlockTime      = time.Now()
		engineAPI          = catalyst.NewConsensusAPI(c.eth)
		header             = c.eth.BlockChain().CurrentHeader()
		curForkchoiceState = engine.ForkchoiceStateV1{
			HeadBlockHash:      header.Hash(),
			SafeBlockHash:      header.Hash(),
			FinalizedBlockHash: header.Hash(),
		}
	)

	// if genesis block, send forkchoiceUpdated to trigger transition to PoS
	if header.Number.BitLen() == 0 {
		if _, err := engineAPI.ForkchoiceUpdatedV2(curForkchoiceState, nil); err != nil {
			log.Crit("failed to initiate PoS transition for genesis via Forkchoiceupdated", "err", err)
		}
	}

	for {
		select {
		case <-c.shutdownCh:
			break
		case curTime := <-ticker.C:
			if curTime.Unix() <= lastBlockTime.Add(c.period).Unix() {
				continue
			}
			c.mu.Lock()
			feeRecipient := c.feeRecipient
			pendingWithdrawals := c.withdrawals.pop()
			c.mu.Unlock()

			if pendingTxs, _ := c.eth.APIBackend.TxPool().Stats(); pendingTxs == 0 && len(pendingWithdrawals) == 0 {
				continue
			}
			// Looks like it's time to build us a block!
			fcState, err := engineAPI.ForkchoiceUpdatedV2(curForkchoiceState, &engine.PayloadAttributes{
				Timestamp:             uint64(curTime.Unix()),
				Random:                common.Hash{}, // TODO: make this configurable?
				SuggestedFeeRecipient: feeRecipient,
				Withdrawals:           pendingWithdrawals,
			})
			if err != nil {
				log.Crit("failed to trigger block building via forkchoiceupdated", "err", err)
			}

			var payload *engine.ExecutableData
			time.Sleep(200 * time.Millisecond) // Give it some time to build
			if payload, err = engineAPI.GetPayloadV1(*fcState.PayloadID); err != nil {
				log.Crit("error retrieving payload", "err", err)
			}
			// Don't accept empty blocks.
			if payload == nil || len(payload.Transactions) == 0 && len(payload.Withdrawals) == 0 {
				continue
			}
			// mark the payload as canon
			if _, err = engineAPI.NewPayloadV2(*payload); err != nil {
				log.Crit("failed to mark payload as canonical", "err", err)
			}
			curForkchoiceState := engine.ForkchoiceStateV1{
				HeadBlockHash:      payload.BlockHash,
				SafeBlockHash:      payload.BlockHash,
				FinalizedBlockHash: payload.BlockHash,
			}
			// mark the block containing the payload as canonical
			if _, err = engineAPI.ForkchoiceUpdatedV2(curForkchoiceState, nil); err != nil {
				log.Crit("failed to mark block as canonical", "err", err)
			}
			lastBlockTime = time.Unix(int64(payload.Timestamp), 0)
		}
	}
}
```

holiman commented on simulated_beacon/simulated_beacon.go:
```diff
@@ -0,0 +1,240 @@
+	)
+
+	// if genesis block, send forkchoiceUpdated to trigger transition to PoS
+	if header.Number.BitLen() == 0 {
```
```suggestion
	if header.Number.Sign() == 0 {
```
A minor nit, which is of no consequence in this case, but generally, `Sign` is better than `BitLen`, since sign just checks the length of `x.abs` whereas `BitLen` actually iterates over the `x.abs`

holiman commented:
Another UX nit: historically, this is how `dev.period` has worked: 
- `0` means "only make blocks if we have transactions". Since block time needs to progress, it basically menas "create blocks as soon as possible, whenever transcations are waiting", typically spits out 1 block per second if there's a steady inflow of transactions
- `non-zero` means "make blocks every N seconds". Regardless of whether transactions are waiting or not. 

What you have implemented is a mixture of the two, where you _both_ require transactions _and_ require the time period to elapse. Nothing wrong with that, but it's a change in semantics, so let's stick with the old semantics.

And with the old semantics, having it default to `0` (blocks on demand) is better than `12`

lightclient commented on simulated_beacon/simulated_beacon.go:
```diff
@@ -0,0 +1,236 @@
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if withdrawal.Index < w.next {
```
I kinda think you should just auto-assign the index. There is technically no index verification on the EL and this already ensures the index in monotonically increasing.

lightclient replied:
seems fine

09306677806 replied:
How do I set the index to automatic? Why is my system set to zero? I have a lot of assets. Why should it be like this? Please guide me?

holiman commented:
Suggestions: https://github.com/ethereum/go-ethereum/commit/d57c9099bb1b1e7a98160d62d9ff50849b854234 (I didn't see your latest commit, might totally clash with that)

jwasinger commented:
@holiman so I may be running out of steam today.  But I think the problem with what you're suggesting is that we don't allow any time between starting building the payload and terminating it with `GetPayload`.

So I afaict, we either need an inner loop to poll after payload has started building.  Or we need a method to poll the in-progress block from the miner.

holiman commented:
See https://github.com/ethereum/go-ethereum/commit/d57c9099bb1b1e7a98160d62d9ff50849b854234#diff-daedd48c27b7dffd18b8d2f8d7dfb600d2ab5f642ab9da2580ac31b7a7f201ebR179

jwasinger commented:
Ah okay.  it seems to work.

holiman commented:
The way it is right now, it _works_, but there's a neater way to do it. Consider the case where the dev.period is 5 seconds. We do this right now

```
for {
 select { 
    every 100 ms: 
        if 5 seconds have elapsed: 
             build_block()
              wait (100 ms) 
             finalize_block()
   } 

}
```
This is inelegant, if we have 5 seconds block time, we could just let the sealer make use of those 5 seconds. So we could structure it like this instead: 
```golang
	fcId = beginSealing()

	for {
		select {
		case <-c.shutdownCh:
			return
		case <-ticker.C:
			if onDemand {
				// Do nothing as long as blocks are empty
				if pendingTxs, _ := c.eth.APIBackend.TxPool().Stats(); pendingTxs == 0 && c.withdrawals.length() == 0 {
					ticker.Reset(buildWaitTime)
					continue
				}
			}
			finalizeSealing(fcId, onDemand)
			fcId = beginSealing()

			if !onDemand {
				ticker.Reset(time.Second * time.Duration(c.period))
			} else {
				ticker.Reset(buildWaitTime)
			}
		}
	}
```
That is, we "always" have a work in progress, and every `block.period` we collect it and schedule a new one. For the `period=0` case, we still need to do some extra legwork.

jwasinger commented:
I like this.  Will try to implement it.

One thing we have to consider is that payloads will time out after 12 seconds.  So for the sake of keeping the logic in this code simple, we should make this timeout configurable.

holiman commented:
> Will try to implement it.

I kind of tried it out already [here](https://github.com/holiman/go-ethereum/commit/1318a654683c1addb5ad459daa47aa6f4203167f), feel free to use or ignore

holiman commented on cmd/geth/simulatedbeacon/simulated_beacon.go:
```diff
@@ -0,0 +1,237 @@
+	}
+	var fcId *engine.PayloadID
+	if fc, err := beginSealing(); err != nil {
+		log.Error("Error starting sealing-work", "err", err)
```
Yes, I think `log.Crit` is usually taking the lazy way out... I find it more elegant to return errors and handle them, rather than log kill the process on failure. 
For example, if a shutdown were in progress, it would be acceptable if the engine rejected a block, due to shuttind down. And in that case, it's not "correct" to just "die". Better to log and exit gracefully, in general, IMO.

holiman commented on cmd/geth/simulatedbeacon/simulated_beacon.go:
```diff
@@ -0,0 +1,240 @@
+			if buildWaitTime < 10*time.Second {
+				buildWaitTime += buildWaitTime
+			}
+			// TODO: don't lose withdrawals
```
This todo can be removed now

jwasinger commented on cmd/geth/simulatedbeacon/simulated_beacon.go:
```diff
@@ -0,0 +1,240 @@
+func RegisterAPIs(stack *node.Node, c *SimulatedBeacon) {
+	stack.RegisterAPIs([]rpc.API{
+		{
+			Namespace: "dev",
```
No there was no discussion.  I'm fine to move it under `debug` and not add a new namespace is this is what we would like.

s1na replied:
I'd vote for adding the `dev` namespace or something like it. Because the plan was at some point to standardize the `debug` namespace. IMO we should at least avoid adding new methods there.

jwasinger commented on cmd/geth/simulatedbeacon/simulated_beacon.go:
```diff
@@ -0,0 +1,240 @@
+		case <-ticker.C:
+			if onDemand {
+				// Do nothing as long as blocks are empty
+				if pendingTxs, _ := c.eth.APIBackend.TxPool().Stats(); pendingTxs == 0 && len(c.pendingWithdrawals) == 0 {
```
Pushed a commit that fixes this.

q9f commented:
Just tested this. Small _nit_ I noticed, we can remove in dev mode:

```
WARN [06-16|12:57:04.708] Beacon client online, but no consensus updates received in a while. Please fix your beacon client to follow the chain! 
WARN [06-16|12:57:54.713] Post-merge network, but no beacon client seen. Please launch one to follow the chain! 
WARN [06-16|13:02:04.744] Beacon client online, but no consensus updates received in a while. Please fix your beacon client to follow the chain! 
WARN [06-16|13:02:54.751] Post-merge network, but no beacon client seen. Please launch one to follow the chain! 
```

holiman commented on cmd/geth/config.go:
```diff
@@ -189,7 +191,21 @@ func makeFullNode(ctx *cli.Context) (*node.Node, ethapi.Backend) {
+		if err != nil {
+			return nil, nil, err
+		}
+
```
```suggestion
```

holiman commented on cmd/geth/simulatedbeacon/simulatedbeacon.go:
```diff
@@ -0,0 +1,257 @@
+	"errors"
+	"sync"
+	"time"
+
```
```suggestion
```
And needs a goimports afterwards

holiman commented on cmd/geth/simulatedbeacon/simulatedbeacon.go:
```diff
@@ -0,0 +1,257 @@
+}
+
+func (c *SimulatedBeacon) getFeeRecipient() common.Address {
+	c.mu.Lock()
+	feeRecipient := c.feeRecipient
+	c.mu.Unlock()
+
+	return feeRecipient
```
```suggestion
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.feeRecipient
```

holiman reviewed (approved):
Some minor formatting-nits, otherwise lgtm

rjl493456442 commented on cmd/geth/simulatedbeacon/api.go:
```diff
@@ -0,0 +1,20 @@
+package simulatedbeacon
```
Please add the license

lightclient commented:
`beaconsim`?

lightclient commented on cmd/geth/simulatedbeacon/simulatedbeacon.go:
```diff
@@ -0,0 +1,254 @@
+}
+
+// getQueued returns withdrawals which are pending inclusion in the next block
+func (w *withdrawalQueue) getQueued() []*types.Withdrawal {
```
```suggestion
func (w *withdrawalQueue) queued() []*types.Withdrawal {
```

lightclient commented on cmd/geth/simulatedbeacon/simulatedbeacon.go:
```diff
@@ -0,0 +1,254 @@
+	c.mu.Unlock()
+}
+
+func (c *SimulatedBeacon) getFeeRecipient() common.Address {
```
```suggestion
func (c *SimulatedBeacon) feeRecipient() common.Address {
```

lightclient commented on cmd/geth/simulatedbeacon/simulatedbeacon.go:
```diff
@@ -0,0 +1,254 @@
+		}
+	}
+
+	beginSealing := func() (engine.ForkChoiceResponse, error) {
```
should these functions be split out?

rjl493456442 replied:
I agree to split them out

rjl493456442 commented on cmd/geth/config.go:
```diff
@@ -189,7 +191,20 @@ func makeFullNode(ctx *cli.Context) (*node.Node, ethapi.Backend) {
 	}
-	return stack, backend
+
+	// Start the dev mode if requested
```
You can also define a function `RegisterBeaconSim` or so in `utils` package and `Fatalf` if any error is occurred.
So that we don't need to return the error.

jwasinger replied:
I modified it to not return an error (it seems that's what the code here already did before, so I guess it makes sense not to change that).  I think it's not really that much code so I didn't put it into its own function.

rjl493456442 commented on cmd/geth/simulatedbeacon/api.go:
```diff
@@ -0,0 +1,36 @@
+}
+
+func (api *API) SetFeeRecipient(ctx context.Context, feeRecipient *common.Address) {
+	api.simBeacon.setFeeRecipient(feeRecipient)
```
Any particular reason to use an address pointer? Maybe `common.Address` is enough I think.

jwasinger replied:
`common.Address` works fine.

rjl493456442 commented on cmd/geth/simulatedbeacon/simulatedbeacon.go:
```diff
@@ -0,0 +1,254 @@
+	return &SimulatedBeacon{
+		eth:          eth,
+		period:       chainConfig.Dev.Period,
+		withdrawals:  withdrawalQueue{[]*types.Withdrawal{}, []*types.Withdrawal{}, sync.Mutex{}},
```
You can simply initialize it by `new(withdrawalQueue)`

rjl493456442 commented on cmd/geth/simulatedbeacon/simulatedbeacon.go:
```diff
@@ -0,0 +1,254 @@
+
+// Start invokes the SimulatedBeacon life-cycle function in a goroutine
+func (c *SimulatedBeacon) Start() error {
+	c.shutdownCh = make(chan struct{})
```
usually we initialize all channels in the constructor, it's a bit uncommon to do it here.

rjl493456442 commented on cmd/geth/simulatedbeacon/simulatedbeacon.go:
```diff
@@ -0,0 +1,254 @@
+// Start invokes the SimulatedBeacon life-cycle function in a goroutine
+func (c *SimulatedBeacon) Start() error {
+	c.shutdownCh = make(chan struct{})
+	go c.loop()
```
You may want to use a waitGroup here to exit the loop elegantly.

rjl493456442 commented on cmd/utils/flags.go:
```diff
@@ -1885,9 +1885,6 @@ func RegisterEthService(stack *node.Node, cfg *ethconfig.Config) (ethapi.Backend
 			Fatalf("Failed to create the LES server: %v", err)
 		}
 	}
-	if err := ethcatalyst.Register(stack, backend); err != nil {
```
Why to remove it?

jwasinger replied:
I move it to [here](https://github.com/ethereum/go-ethereum/pull/27327/files#diff-efb34882aeb5938c3c2527733e914b31697acefcdccc3b3c42467fda64efc362R196-R205) .

The `SimulatedBeacon` uses its own instance of `ConsensusAPI`.  I avoid instantiating it twice because it opens the auth RPC API port by default and causes some useless log warnings (https://github.com/ethereum/go-ethereum/pull/27327#issuecomment-1594507869).

rjl493456442 commented on cmd/geth/simulatedbeacon/api.go:
```diff
@@ -0,0 +1,36 @@
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package simulatedbeacon
```
Usually `cmd/geth` is the package we generate the main executable binary. It's weird to put a package here.
I think we can put it inside of `beacon` package.

jwasinger replied:
I put it under `catalyst` for now.

jwasinger commented on eth/catalyst/simulated_beacon.go:
```diff
@@ -0,0 +1,244 @@
+	period       uint64
+	withdrawals  withdrawalQueue
+	// the fee recipient
+	feeTarget common.Address
```
because I renamed the accessor function from `getFeeRecipient` to `feeRecipient`.

lightclient replied:
hmm, if it is only accessed internally why not just access the var directly instead of a function?

jwasinger replied:
So the getter function was used because access to the variable is gated by a mutex.  However, it's only called once so I just removed the getter and renamed the variable to `feeRecipient`.

holiman commented:
Segfault on appveyor: https://ci.appveyor.com/project/ethereum/go-ethereum/builds/47354481/job/yw77wmyflvqx5luk#L943 , via `TestPriorityClient`

rjl493456442 commented on eth/catalyst/simulated_beacon.go:
```diff
@@ -0,0 +1,249 @@
+
+// loop manages the lifecycle of the SimulatedBeacon.
+// it drives block production, taking the role of a CL client and interacting with Geth via public engine/eth APIs
+func (c *SimulatedBeacon) loop() {
```
I think we can split this big loop into two parts: on demand or regular, and also some fields can be initialized in the constructor. Please take a look.

```diff
diff --git a/eth/catalyst/simulated_beacon.go b/eth/catalyst/simulated_beacon.go
index b10ce497de..3e4a58657c 100644
--- a/eth/catalyst/simulated_beacon.go
+++ b/eth/catalyst/simulated_beacon.go
@@ -24,6 +24,7 @@ import (
 
 	"github.com/ethereum/go-ethereum/beacon/engine"
 	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/eth"
 	"github.com/ethereum/go-ethereum/log"
@@ -79,13 +80,14 @@ func (w *withdrawalQueue) clearQueued() {
 }
 
 type SimulatedBeacon struct {
-	shutdownCh   chan struct{}
-	eth          *eth.Ethereum
-	period       uint64
-	withdrawals  withdrawalQueue
-	feeRecipient common.Address
-	// mu gates concurrent access to the feeRecipient
-	mu                 sync.Mutex
+	shutdownCh  chan struct{}
+	eth         *eth.Ethereum
+	period      uint64
+	withdrawals withdrawalQueue
+
+	feeRecipient     common.Address
+	feeRecipientLock sync.Mutex // lock gates concurrent access to the feeRecipient
+
 	engineAPI          *ConsensusAPI
 	curForkchoiceState engine.ForkchoiceStateV1
 	buildWaitTime      time.Duration
@@ -97,24 +99,44 @@ func NewSimulatedBeacon(eth *eth.Ethereum) (*SimulatedBeacon, error) {
 	if chainConfig.Dev == nil {
 		return nil, errors.New("incompatible pre-existing chain configuration")
 	}
+	header := eth.BlockChain().CurrentHeader()
+	current := engine.ForkchoiceStateV1{
+		HeadBlockHash:      header.Hash(),
+		SafeBlockHash:      header.Hash(),
+		FinalizedBlockHash: header.Hash(),
+	}
+	engineAPI := NewConsensusAPI(eth)
 
+	// if genesis block, send forkchoiceUpdated to trigger transition to PoS
+	if header.Number.Sign() == 0 {
+		if _, err := engineAPI.ForkchoiceUpdatedV2(current, nil); err != nil {
+			return nil, err
+		}
+	}
 	return &SimulatedBeacon{
-		eth:           eth,
-		period:        chainConfig.Dev.Period,
-		shutdownCh:    make(chan struct{}),
-		buildWaitTime: time.Millisecond * 100,
+		eth:                eth,
+		period:             chainConfig.Dev.Period,
+		shutdownCh:         make(chan struct{}),
+		buildWaitTime:      time.Millisecond * 100,
+		engineAPI:          engineAPI,
+		lastBlockTime:      header.Time,
+		curForkchoiceState: current,
 	}, nil
 }
 
 func (c *SimulatedBeacon) setFeeRecipient(feeRecipient common.Address) {
-	c.mu.Lock()
+	c.feeRecipientLock.Lock()
 	c.feeRecipient = feeRecipient
-	c.mu.Unlock()
+	c.feeRecipientLock.Unlock()
 }
 
 // Start invokes the SimulatedBeacon life-cycle function in a goroutine
 func (c *SimulatedBeacon) Start() error {
-	go c.loop()
+	if c.period == 0 {
+		go c.loopOnDemand()
+	} else {
+		go c.loop()
+	}
 	return nil
 }
 
@@ -130,9 +152,9 @@ func (c *SimulatedBeacon) beginSealing() (engine.ForkChoiceResponse, error) {
 	if tstamp <= c.lastBlockTime {
 		tstamp = c.lastBlockTime + 1
 	}
-	c.mu.Lock()
+	c.feeRecipientLock.Lock()
 	feeRecipient := c.feeRecipient
-	c.mu.Unlock()
+	c.feeRecipientLock.Unlock()
 
 	return c.engineAPI.ForkchoiceUpdatedV2(c.curForkchoiceState, &engine.PayloadAttributes{
 		Timestamp:             tstamp,
@@ -147,7 +169,6 @@ func (c *SimulatedBeacon) finalizeSealing(id *engine.PayloadID, onDemand bool) e
 	if err != nil {
 		return fmt.Errorf("error retrieving payload: %v", err)
 	}
-
 	if onDemand && len(payload.Transactions) == 0 && len(payload.Withdrawals) == 0 {
 		// If the payload is empty, despite there being pending transactions,
 		// that indicates that we need to give it more time to build the block.
@@ -175,65 +196,53 @@ func (c *SimulatedBeacon) finalizeSealing(id *engine.PayloadID, onDemand bool) e
 	return nil
 }
 
-// loop manages the lifecycle of the SimulatedBeacon.
-// it drives block production, taking the role of a CL client and interacting with Geth via public engine/eth APIs
-func (c *SimulatedBeacon) loop() {
+func (c *SimulatedBeacon) loopOnDemand() {
 	var (
-		ticker = time.NewTimer(time.Second * time.Duration(c.period))
-		header = c.eth.BlockChain().CurrentHeader()
+		newTxs = make(chan core.NewTxsEvent)
+		sub    = c.eth.TxPool().SubscribeNewTxsEvent(newTxs)
 	)
+	defer sub.Unsubscribe()
 
-	c.curForkchoiceState = engine.ForkchoiceStateV1{
-		HeadBlockHash:      header.Hash(),
-		SafeBlockHash:      header.Hash(),
-		FinalizedBlockHash: header.Hash(),
-	}
-	c.engineAPI = NewConsensusAPI(c.eth)
-	c.lastBlockTime = header.Time
-
-	// if genesis block, send forkchoiceUpdated to trigger transition to PoS
-	if header.Number.Sign() == 0 {
-		if _, err := c.engineAPI.ForkchoiceUpdatedV2(c.curForkchoiceState, nil); err != nil {
-			log.Error("failed to initiate PoS transition for genesis via Forkchoiceupdated", "err", err)
+	for {
+		select {
+		case <-c.shutdownCh:
 			return
+		case <-newTxs:
+			resp, err := c.beginSealing()
+			if err != nil {
+				log.Error("Error starting sealing-work", "err", err)
+				return
+			}
+			time.Sleep(c.buildWaitTime)
+
+			err = c.finalizeSealing(resp.PayloadID, true)
+			if err != nil {
+				log.Error("Error collecting sealing-work", "err", err)
+				return
+			}
 		}
 	}
+}
 
-	var fcId *engine.PayloadID
-	if fc, err := c.beginSealing(); err != nil {
-		log.Error("Error starting sealing-work", "err", err)
-		return
-	} else {
-		fcId = fc.PayloadID
-	}
-	onDemand := (c.period == 0)
+// loop manages the lifecycle of the SimulatedBeacon.
+// it drives block production, taking the role of a CL client and interacting with Geth via public engine/eth APIs
+func (c *SimulatedBeacon) loop() {
+	timer := time.NewTimer(time.Second * time.Duration(c.period))
 	for {
+		resp, err := c.beginSealing()
+		if err != nil {
+			log.Error("Error starting sealing-work", "err", err)
+			return
+		}
 		select {
 		case <-c.shutdownCh:
 			return
-		case <-ticker.C:
-			if onDemand {
-				// Do nothing as long as blocks are empty
-				if pendingTxs, _ := c.eth.APIBackend.TxPool().Stats(); pendingTxs == 0 && len(c.withdrawals.queued()) == 0 {
-					ticker.Reset(c.buildWaitTime)
-					continue
-				}
-			}
-			if err := c.finalizeSealing(fcId, onDemand); err != nil {
+		case <-timer.C:
+			if err := c.finalizeSealing(resp.PayloadID, false); err != nil {
 				log.Error("Error collecting sealing-work", "err", err)
 				return
 			}
-			if fc, err := c.beginSealing(); err != nil {
-				log.Error("Error starting sealing-work", "err", err)
-				return
-			} else {
-				fcId = fc.PayloadID
-			}
-			if !onDemand {
-				ticker.Reset(time.Second * time.Duration(c.period))
-			} else {
-				ticker.Reset(c.buildWaitTime)
-			}
+			timer.Reset(time.Second * time.Duration(c.period))
 		}
 	}
 }

```

jwasinger replied:
One issue with this is that if we are in on-demand mode and the payload takes longer to build than `buildWaitTime`, the block will not be built.

rjl493456442 replied:
Yup, but you have the mechanism to double the wait time. So the wait time will gradually climb to a proper value, isn't it?

This logic is not changed.

jwasinger replied:
I should have elaborated...  the issue is that in `loopOnDemand` the build is triggered by the `NewTxs` event.  So if a block takes too long to build, the build won't re-trigger until there are new txs in the pool.

rjl493456442 replied:
We can use a new API to retrieve payload in sync mode. https://github.com/rjl493456442/go-ethereum/commit/9f467bccd130d77efe2a48026f64930cfe5c0f5e

But yeah, if something weird happens(we have an executable transaction but later on it's gone somehow) then GetFullPayload will still return empty-one, but I think it's fine to dev mode.

jwasinger replied:
If we receive a new transaction in the pool, the build triggers, and the transaction takes longer to build than the default time, the build loop won't re-trigger until there are other transactions in the pool.

rjl493456442 replied:
No, you don't need the wait mechanism if you get payload in blocked way.

jwasinger commented on eth/catalyst/simulated_beacon.go:
```diff
@@ -0,0 +1,249 @@
+	if _, err = c.engineAPI.ForkchoiceUpdatedV2(c.curForkchoiceState, nil); err != nil {
+		return fmt.Errorf("failed to mark block as canonical: %v", err)
+	}
+	c.withdrawals.clearQueued()
```
good catch!

jwasinger commented on eth/catalyst/simulated_beacon.go:
```diff
@@ -0,0 +1,249 @@
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	w.pending = append(w.pending, withdrawal)
```
Is 20 withdrawals a parameter used by the beacon chain?  Or is this just to provide a sane limit on how big the queue can/should get?

rjl493456442 replied:
Just an arbitrary number...

jwasinger commented on eth/catalyst/simulated_beacon.go:
```diff
@@ -0,0 +1,249 @@
+}
+
+func NewSimulatedBeacon(eth *eth.Ethereum) (*SimulatedBeacon, error) {
+	chainConfig := eth.APIBackend.ChainConfig()
```
I changed flag type to Uint64Flag which ensures proper behavior on each of these points.

lightclient commented on eth/catalyst/simulated_beacon.go:
```diff
@@ -0,0 +1,247 @@
+}
+
+// loop manages the lifecycle of the SimulatedBeacon.
+// it drives block production, taking the role of a CL client and interacting with Geth via public engine/eth APIs
```
```suggestion
// It drives block production, taking the role of a CL client and interacting with Geth via public engine/eth APIs.
```

lightclient commented on eth/catalyst/simulated_beacon.go:
```diff
@@ -0,0 +1,247 @@
+		queueCount = len(w.pending)
+	}
+
+	p := make([]*types.Withdrawal, queueCount)
```
Can replace the above with:
```suggestion
        count := min(10, len(w.pending))
	p := make([]*types.Withdrawal, count)
```

jwasinger replied:
`min` isn't a function though?

lightclient replied:
I think it would be nice. They're being added to go 1.21 so we can always remove it after a few more releases.

jwasinger replied:
This is no longer relevant after the refactor I pushed today.

jwasinger commented on eth/catalyst/simulated_beacon.go:
```diff
@@ -0,0 +1,238 @@
+	}
+
+	// TODO: this is a bit of a hack.  need to modify payload builder to stop payload w/ GetFullPayload
+	_, _ = c.engineAPI.GetPayloadV1(*fcResponse.PayloadID)
```
If I don't explicitly stop building the payload with this call here, it will continue building (even after it is included in the chain).  So maybe this should be handled within the payload building code in the miner instead of a hack call here.  Will look into it.

jwasinger commented on eth/catalyst/simulated_beacon.go:
```diff
@@ -0,0 +1,238 @@
+
+// withdrawalQueue implements a FIFO queue which holds withdrawals that are pending inclusion
+type withdrawalQueue struct {
+	pending chan *types.Withdrawal
```
I'm unsure if there is a reason to explicitly close this channel on client shutdown.  I don't now because it could cause a panic if the user is making `dev_addWithdrawal` calls at the moment when the client shuts down.

jwasinger commented:
@rjl493456442 I think this is in a good state for another round of reviews now (other than the issue I mention about the payload building).  IMO the miner changes should probably be broken into their own PR to keep this PR more compact.

rjl493456442 reviewed (approved):
LGTM!

holiman commented on eth/catalyst/simulated_beacon.go:
```diff
@@ -0,0 +1,231 @@
+	feeRecipient := c.feeRecipient
+	c.feeRecipientLock.Unlock()
+
+	fcResponse, err := c.engineAPI.ForkchoiceUpdatedV2(c.curForkchoiceState, &engine.PayloadAttributes{
+		Timestamp:             tstamp,
+		SuggestedFeeRecipient: feeRecipient,
+		Withdrawals:           withdrawals,
+	})
+	if err != nil {
+		return fmt.Errorf("error calling forkchoice update: %v", err)
+	}
+
+	envelope, err := c.engineAPI.getFullPayload(*fcResponse.PayloadID)
+	if err != nil {
+		return fmt.Errorf("error retrieving payload: %v", err)
+	}
+	payload := envelope.ExecutionPayload
```
I haven't been following the developments lately. A while back, it was designed so that there was always a block in production, and new transactions would be added to it, and whenever we wanted to, we could collect it. Meaning that the sealer has a lot of time to seal it: if the time between blocks is 10 seconds, the prior design would have 10 seconds to work on each block. 

How is it now? It _looks_ to me like we are giving the miner zero time to work on the block, before we collect it. Am I missing something?

jwasinger replied:
With the new API, if there are transactions in the pool `getFullPayload` will hang until a full block is created so we no longer need delay logic within the `SimulatedBeacon`.

holiman reviewed (approved):
LGTM

hadv commented:
Hi, 

>   3. A random, pre-allocated developer account will be available and unlocked as
     eth.coinbase, which can be used for testing. The random dev account is temporary,
     stored on a ramdisk, and will be lost if your machine is restarted.
     
   Using random account for testing is not a good idea, especially with automation test environment. 
   Can you add a flag to pre-allocate fund for specific accounts?

jwasinger commented:
@hadv if you want to fund specific accounts, transfer ether from the coinbase account.  There isn't a need for the ability to pre-allocate specific accounts at genesis.

hadv commented:
> @hadv if you want to fund specific accounts, transfer ether from the coinbase account. There isn't a need for the ability to pre-allocate specific accounts at genesis.

When using **web3j** for automation testing, we need Credentials to send fund; but we didn't have Credentials of coinbase account

rmlearney-digicatapult commented:
I've just opened issue #28449 related to this issue - is it possible to run a private Shanghai-EVM-compatible PoA network without `--dev`?
